crates/ide-diagnostics/src/handlers/trait_impl_redundant_assoc_item.rs

```diff
@@ -10,47 +17,196 @@ pub(crate) fn trait_impl_redundant_assoc_item(
     ctx: &DiagnosticsContext<'_>,
     d: &hir::TraitImplRedundantAssocItems,
 ) -> Diagnostic {
+    let db = ctx.sema.db;
     let name = d.assoc_item.0.clone();
+    let redundant_assoc_item_name = name.display(db);
     let assoc_item = d.assoc_item.1;
-    let db = ctx.sema.db;
 
     let default_range = d.impl_.syntax_node_ptr().text_range();
     let trait_name = d.trait_.name(db).to_smol_str();
 
-    let (redundant_item_name, diagnostic_range) = match assoc_item {
-        hir::AssocItem::Function(id) => (
-            format!("`fn {}`", name.display(db)),
-            Function::from(id)
-                .source(db)
-                .map(|it| it.syntax().value.text_range())
-                .unwrap_or(default_range),
-        ),
-        hir::AssocItem::Const(id) => (
-            format!("`const {}`", name.display(db)),
-            Const::from(id)
-                .source(db)
-                .map(|it| it.syntax().value.text_range())
-                .unwrap_or(default_range),
-        ),
-        hir::AssocItem::TypeAlias(id) => (
-            format!("`type {}`", name.display(db)),
-            TypeAlias::from(id)
-                .source(db)
-                .map(|it| it.syntax().value.text_range())
-                .unwrap_or(default_range),
-        ),
+    let (redundant_item_name, diagnostic_range, redundant_item_def) = match assoc_item {
+        hir::AssocItem::Function(id) => {
+            let function = Function::from(id);
+            (
+                format!("`fn {}`", redundant_assoc_item_name),
+                function
+                    .source(db)
+                    .map(|it| it.syntax().value.text_range())
+                    .unwrap_or(default_range),
+                format!("\n    {};", function.display(db).to_string()),
+            )
+        }
+        hir::AssocItem::Const(id) => {
+            let constant = Const::from(id);
+            (
+                format!("`const {}`", redundant_assoc_item_name),
+                constant
+                    .source(db)
+                    .map(|it| it.syntax().value.text_range())
+                    .unwrap_or(default_range),
+                format!("\n    {};", constant.display(db).to_string()),
+            )
+        }
+        hir::AssocItem::TypeAlias(id) => {
+            let type_alias = TypeAlias::from(id);
+            (
+                format!("`type {}`", redundant_assoc_item_name),
+                type_alias
+                    .source(db)
+                    .map(|it| it.syntax().value.text_range())
+                    .unwrap_or(default_range),
+                format!("\n    type {};", type_alias.name(ctx.sema.db).to_smol_str()),
+            )
+        }
     };
 
     Diagnostic::new(
         DiagnosticCode::RustcHardError("E0407"),
         format!("{redundant_item_name} is not a member of trait `{trait_name}`"),
         FileRange { file_id: d.file_id.file_id().unwrap(), range: diagnostic_range },
```

Comment:
Oh I just realized I missed this on the review that added this diagnostic, but this unwrap is very bad 😅 We should be using `adjusted_display_range_new` to get the new range for this here.

Reply:
`adjusted_display_range_new` require a `diag_ptr: InFile<AstPtr<N>>`, regarding get `diag_ptr` from the current `hir::AssocItem`, we could use  `hir::AssocItem.source(db)`, which return an `Option<InFile<Self::Ast>>`, could we `unwarp` or `expect` directly? since `hir::AssocItem.source(db)` actually return `Some(...)`

Reply:
Can also do something similar to `adjusted_display_range_new` since we are skipping the `diag_ptr`. And no unwrapping isn't an option here, the source fetching is technically fully infallible right now, but we might change things at some point where we allow loading libraries with missing sources at which point the unwrap would fail.

Reply:
Wasn't aware this was already discussed here. Opened a issue #16269.